### PR TITLE
MM-24335 - show recently viewed channels

### PR DESCRIPTION
**Pull request by sridhar02**
#### Summary
To show recently viewed channels when channel switcher opens #14633
#### Ticket Link

  Fixes https://github.com/mattermost/mattermost-server/issues/14633

#### Related Pull Requests
- Has redux changes  https://github.com/mattermost/mattermost-redux/pull/1201
#### Screenshots
image

### CODE REVIEW

**Comment by sridhar02** [comment]
I have some doubts regarding this issue 
1.whether to change the commit SHA number of matter most-redux to the latest commit SHA in package.json
2.do I need to change all the localization files to show the recent channels in their respective languages or is there any alternative to generate the languages
3. I think this PR is not yet done I am not showing the unread channels in the list and some other are also messed up I guess, can anyone review this PR and tell me what is to addressed and changed in this PR.

**Comment by esethna** [comment]
@hmhealey can you help with guidance here ^

**Comment by hmhealey** [comment]
> 1.whether to change the commit SHA number of matter most-redux to the latest commit SHA in package.json

As long as you update your PR to be off the latest master, you don't need to make any changes to the package.json. Someone else has already updated it to a newer commit of mattermost-redux, so it already includes those changes.

> 2.do I need to change all the localization files to show the recent channels in their respective languages or is there any alternative to generate the languages

No, you don't need to make any other changes. The translation tools just detect the strings in the code and in en.json, and it uses those to tell the translators that we need the other versions.

There is a failure in the i18n-check job [here](https://app.circleci.com/pipelines/github/mattermost/mattermost-webapp/23263/workflows/006e0647-597d-4633-8d78-26563b29aa17/jobs/148895), but I think that might be because you added the new string out of order. It should come before `suggestion.mention.special` alphabetically.

3. I think this PR is not yet done I am not showing the unread channels in the list and some other are also messed up I guess, can anyone review this PR and tell me what is to addressed and changed in this PR.

It sounds like we want to show recent channels instead of unread channels before the user has typed anything in, so maybe the call to `this.formatUnreadChannelsAndDispatch` can be removed. Maybe that's also what's causing the other items to be messed up.

**Comment by sridhar02** [comment]
@hmhealey can you review the PR, I have made necessary changes for the PR

File: components/suggestion/switch_channel_provider.jsx
Comment (ali-farooq0) [inline components/suggestion/switch_channel_provider.jsx:532]: This can be simplified to:

```suggestion
        for (const channel of recentChannels) {
```

File: components/suggestion/switch_channel_provider.jsx
Comment (ali-farooq0) [inline components/suggestion/switch_channel_provider.jsx:556]: should this be removed?

File: components/suggestion/switch_channel_provider.jsx
Comment (ali-farooq0) [inline components/suggestion/switch_channel_provider.jsx:554]: hmm why is this `channel.id` if we're returning `channelNames`? does it get resolved somewhere?

File: components/suggestion/switch_channel_provider.jsx
Comment (ali-farooq0) [inline components/suggestion/switch_channel_provider.jsx:286]: should `formatUnreadChannelsAndDispatch` function be removed if it's not being used?

**Review by ali-farooq0 — CHANGES_REQUESTED** [review]
Thanks for your contribution @sridhar02! Just a few comments/concerns.

**Comment by esethna** [comment]
Thanks @sridhar02, however I don't think this is working as expected. Notice in this GIF, I would expect "Test 1" to be at the top, given I just viewed that channel:

https://i.gyazo.com/1908f68e1f1f6ef7f35755ea64739f34.gif

**Comment by sridhar02** [comment]
@esethna I will look into it,I having other work was not able work on it. I will complete this one in a week or two.

**Comment by esethna** [comment]
Hi @sridhar02, let us know if you're still working on this issue or need help, thanks!

File: components/suggestion/switch_channel_provider.jsx
Comment (sridhar02) [reply components/suggestion/switch_channel_provider.jsx:548]: applied this change

File: components/suggestion/switch_channel_provider.jsx
Comment (sridhar02) [reply components/suggestion/switch_channel_provider.jsx:572]: Removed it.

File: components/suggestion/switch_channel_provider.jsx
Comment (sridhar02) [reply components/suggestion/switch_channel_provider.jsx:286]: I think I removed this one from the code.

File: components/suggestion/switch_channel_provider.jsx
Comment (sridhar02) [reply components/suggestion/switch_channel_provider.jsx:570]: I wrote this using the seeing the above function of `formatUnreadChannelsAndDispatch` and I have not looked why it is written that way.I will check it.

**Comment by sridhar02** [comment]
@ali-farooq0 I have updated with the changes can you review it once more, as esethna mentioned I was not able to see the recent channels in the switch modal, can you help why I are getting this issue ?

**Comment by esethna** [comment]
@sridhar02 let us know when this is ready to re-review, seeing the same issue I noted last time ^ https://github.com/mattermost/mattermost-webapp/pull/6198#issuecomment-686669167

**Comment by ali-farooq0** [comment]
Going to hold off on reviewing this. Let us know when this is ready for review @sridhar02. Thanks!

**Comment by sridhar02** [comment]
@hmhealey can you help with the PR as observed it is not showing recent channels in the quick channel switcher, as I am getting data from the redux selector is there any problems with the selector or any other issue I am not able to identify it.

**Comment by hmhealey** [comment]
Oh, I see what the problem is. We're using `getAllRecentChannels` to populate the quick switcher, but "recent" in that case refers to the time of the last post in the channel, not the last time the current user viewed the channel. We really need to be more careful when we use "recent" for different things. Sorry about the confusion here. I didn't realize the selector you wrote didn't match up with what this ticket wanted until now.

So this is going to require a new selector, or at least changes to the previous one. I think the channel membership object for each channel keeps track of the `last_viewed_at` for each channel, and we can use that here. The only potential problem with that is that it isn't updated immediately when you view a channel since it also drives the New Messages line in the post list.

For the new selector, you can also make that in this repo instead of worrying about submitting another PR to mattermost-redux.

**Comment by esethna** [comment]
@sridhar02 how's this going? Let us know when this is ready for re-review

**Comment by jfrerich** [comment]
Hi, @sridhar02! Just checking in to see if you need any help getting your PR prepped for another round of reviews.  Thanks for the hard work!

**Comment by sridhar02** [comment]
@jfrerich I am not able to complete this, can you reopen this to someone, I am not getting time to complete this issue?

**Comment by esethna** [comment]
@hmhealey do we know how close this is to completion? Should be bring it internally or reopen the issue?

**Comment by jfrerich** [comment]
@sridhar02, that's not a problem and understandable and I appreciate you letting us know!  I'll leave this to @esethna and @hmhealey to determine the next steps

**Comment by esethna** [comment]
cc// @hmhealey on this one, I'm not sure how far this is from completion? If it's not far off perhaps we can get @asaadmahmood's help with guidance from you

**Comment by hmhealey** [comment]
The redux code need to be redone since this was one of the PRs where we had confused recently read channels with recently viewed ones, so it's not simple to finish this. I think our best bet is to just close this for now to let someone else take a shot at it.

**Pull request closed without merging**
